clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -2048,10 +2050,37 @@ void CodeGenModule::SetLLVMFunctionAttributesForDefinition(const Decl *D,
       B.addAttribute(llvm::Attribute::MinSize);
   }
 
-  if (D->hasAttr<YkOutlineAttr>()) {
+  bool Outline = false;
+  if (D->hasAttr<YkOutlineAttr>())
+    Outline = true;
+
+  // Mark all functions containing loops with `yk_outline` unless the function
+  // was annotated `yk_unroll_safe`.
+  //
+  // If `F->empty()` then this is merely a function declaration for which we
+  // have no IR. In this case the JIT will be unable to inline calls to this
+  // function anyway, so there's no need to conservatively add `yk_outline`.
+  if (!F->empty()) {
+    llvm::DominatorTree DT(*F);
+    llvm::LoopInfo LI(DT);
+    if (!LI.empty()) {
+      if (D->hasAttr<YkUnrollSafeAttr>()) {
+        // It is safe to inline `F` despite it containing loops. We also mark
+        // it `noinline` to block inlining during AOT compilation. Without this
+        // `F` may get AOT inlined into a parent function which contains loops
+        // and isn't marked `yk_unroll_safe`. In this case `F` would not in
+        // fact be inlined into the trace!
+        B.addAttribute(llvm::Attribute::NoInline);
+      } else {
+        Outline = true;
```

Comment:
Since `if (Outline)` on 2080 adds `NoInline` we're actually adding `NoInline` in both the true and false branches of this `if`! As this might suggest, I think the `Outline` variable is making the code harder to read and we'd probably be best off without it.

I would prefer something simple such as:

```
if (D->hasAttr<YkUnrollSafeAttr>()) {
    B.addAttribute("yk_outline");
    B.addAttribute(llvm::Attribute::NoInline);
}
```

Reply:
Try [821d36b](https://github.com/ykjit/ykllvm/pull/44/commits/821d36b75b5f6390b9742c3361bfe98f08a22c2f)

I've resisted lifting `B.addAttribute(llvm::Attribute::NoInline);` out of the conditional for de-duplication, since doing so makes the code harder to follow IMHO. I hope you agree.

Reply:
I'm confused. Why is:

```
if (X) {
  Y
} else {
  Y
  Z
}
```

something I would ever want to write? It's always much clearer as:

```
Y
if (!X) {
  Z
}
```

In our case, we're marking *all* functions with loops as `NoInline` whether or not they're unroll safe. That doesn't seem right.

Reply:
Does 5c95ec4ad3aa make sense?

I must have re-written this comment 5 or 6 times and I'm not totally happy with what I've got here either.

We need to explain why we `noinline` all loopy functions. It has to do with "parent pollution", but for subtly different reasons depending upon if the function is `unroll_safe` or not.